communication/nats/connection_wrap.go

```diff
@@ -109,19 +123,42 @@ func (c *ConnectionWrap) connectOptions() nats_lib.Options {
 	return options
 }
 
-// Open starts the connection: left for test compatibility.
+func (c *ConnectionWrap) resurrect() {
+	for {
+		select {
+		case <-c.closeCh:
+			return
+		case <-time.After(ReconnectWait):
+			err := c.open()
+			if err == nil {
+				return
+			}
+		}
+	}
+}
+
+// open starts the connection: left for test compatibility.
 // Deprecated: Use nats.BrokerConnector#Connect() instead.
-func (c *ConnectionWrap) Open() (err error) {
-	c.Conn, err = c.connectOptions().Connect()
+func (c *ConnectionWrap) open() (err error) {
+	newConn, err := c.connectOptions().Connect()
 	if err != nil {
 		log.Warn().Err(err).Msgf("Failed to connect to NATS servers %v, will reconnect again", c.connectOptions().Servers)
 	}
 
+	c.connMux.Lock()
+	defer c.connMux.Unlock()
+	c.Conn = newConn
+
 	return nil
 }
 
 // Close destructs the connection.
 func (c *ConnectionWrap) Close() {
+	c.closeOnce.Do(c.close)
+}
+
+func (c *ConnectionWrap) close() {
+	close(c.closeCh)
 	if c.Conn != nil {
```

Comment:
Am I reading this wrong or should add a mutex here too to avoid the race detector yelling?

Reply:
ConnectionWrap.close() invocation from ConnectionWrap.Close() is guarded by sync.Once, therefore no races possible